### PR TITLE
[pulsar-broker] Stop to dispatch when skip message temporally since Key_Shared consumer stuck on delivery

### PR DESCRIPTION
**Pull request by equanz**
### Motivation
In some case of Key_Shared consumer, messages ordering was broken.
Here is how to reproduce(I think it is one of case to reproduce this issue).

1. Connect Consumer1 to Key_Shared subscription `sub` and stop to receive
   - receiverQueueSize: 500
2. Connect Producer and publish 500 messages with key `(i % 10)`
3. Connect Consumer2 to same subscription and start to receive
   - receiverQueueSize: 1
   - since https://github.com/apache/pulsar/pull/7106 , Consumer2 can't receive (expected)
4. Producer publish more 500 messages with same key generation algorithm
5. After that, Consumer1 start to receive
6. Check Consumer2 message ordering
   - sometimes message ordering was broken in same key

Consumer1:
```
Connected: Tue Jul 14 09:36:39 JST 2020
[pulsar-client-io-1-1] WARN com.scurrilous.circe.checksum.Crc32cIntChecksum - Failed to load Circe JNI library. Falling back to Java based CRC32c provider
[pulsar-timer-4-1] INFO org.apache.pulsar.client.impl.ConsumerStatsRecorderImpl - [persistent://public/default/key-shared-test] [sub0] [820f0] Prefetched messages: 499 --- Consume throughput received: 0.02 msgs/s --- 0.00 Mbit/s --- Ack sent rate: 0.00 ack/s --- Failed messages: 0 --- batch messages: 0 ---Failed acks: 0
Received: my-message-0 PublishTime: 1594687006203 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-1 PublishTime: 1594687006243 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-2 PublishTime: 1594687006247 Date: Tue Jul 14 09:37:46 JST 2020
...
Received: my-message-498 PublishTime: 1594687008727 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-499 PublishTime: 1594687008731 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-500 PublishTime: 1594687038742 Date: Tue Jul 14 09:37:46 JST 2020
...
Received: my-message-990 PublishTime: 1594687040094 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-994 PublishTime: 1594687040103 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-995 PublishTime: 1594687040105 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-997 PublishTime: 1594687040113 Date: Tue Jul 14 09:37:46 JST 2020
```

Consumer2:
```
Connected: Tue Jul 14 09:37:03 JST 2020
[pulsar-client-io-1-1] WARN com.scurrilous.circe.checksum.Crc32cIntChecksum - Failed to load Circe JNI library. Falling back to Java based CRC32c provider
Received: my-message-501 MessageId: 4:1501:-1 PublishTime: 1594687038753 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-502 MessageId: 4:1502:-1 PublishTime: 1594687038755 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-503 MessageId: 4:1503:-1 PublishTime: 1594687038759 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-506 MessageId: 4:1506:-1 PublishTime: 1594687038785 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-508 MessageId: 4:1508:-1 PublishTime: 1594687038812 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-901 MessageId: 4:1901:-1 PublishTime: 1594687039871 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-509 MessageId: 4:1509:-1 PublishTime: 1594687038815 Date: Tue Jul 14 09:37:46 JST 2020
ordering was broken, key: 1 oldNum: 901 newNum: 511
Received: my-message-511 MessageId: 4:1511:-1 PublishTime: 1594687038826 Date: Tue Jul 14 09:37:46 JST 2020
Received: my-message-512 MessageId: 4:1512:-1 PublishTime: 1594687038830 Date: Tue Jul 14 09:37:46 JST 2020
...
```

I think this issue is caused by https://github.com/apache/pulsar/pull/7105.
Here is an example.
1. dispatch messages
2. Consumer2 was stuck and `totalMessagesSent=0`
   - Consumer2 availablePermits was 0
3. skip redeliver messages temporally
   - Consumer2 availablePermits was back to 1
4. dispatch new messages
   - new message was dispatched to Consumer2
5. back to redeliver messages
4. dispatch messages
   - ordering was broken

### Modifications
Stop to dispatch when skip message temporally since Key_Shared consumer stuck on delivery.

### Verifying this change
- [x] Make sure that the change passes the CI checks.

It seems that this issue should check in some test cases.
However, I think this issue is corner case and not easier to check. If it is not, please tell me.

### Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation
  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by feeblefakie** [comment]
I've check the procedure described in #7455 with equanz:stop_to_dispatch_manually_key_shared.
This time, consumers got to not consuming any messages now. Seems just stuck.

**Comment by feeblefakie** [comment]
I've checked this PR with the procedure in #7455 to see if there are any out-of-order messages.
But now consumers don't even consume anything.
The same thing happens for pulsar-perf so I don't think it's the issue of the procedure.

**Comment by equanz** [comment]
@feeblefakie Thank you for your check. But I couldn't reproduce your issue (my procedure below) now. Is [this debug log](https://github.com/apache/pulsar/pull/7553/files#diff-8f773b51748a5cb5f7a298a4f65d1134R264) occurred in your standalone? Or please tell me more detail.

test procedure
1. build this code
2. start standalone with `--wipe-data` and default config (so, use `AUTO_SPLIT`)
3. create partitioned topic with 32 partition
4. create subscription `sub0`, and `sub1` as latest position to same topic
5. produce messages with your [code](https://github.com/feeblefakie/misc/blob/ed6439c596227eace6ce7176d93424fa35d34406/pulsar/src/main/java/MyProducer.java)
6. consume messages as `sub0` with your [code](https://github.com/feeblefakie/misc/blob/ed6439c596227eace6ce7176d93424fa35d34406/pulsar/src/main/java/MyConsumer.java)
   - out: 10000 records
7. consume messages as `sub1` with your [code](https://github.com/feeblefakie/misc/blob/ed6439c596227eace6ce7176d93424fa35d34406/pulsar/src/main/java/MyConsumer.java)
   - out: 10000 records
8. check output with your [script](https://github.com/feeblefakie/misc/blob/ed6439c596227eace6ce7176d93424fa35d34406/pulsar/check.sh)

and no `NG` was found(with 1000 `OK`).

**Comment by feeblefakie** [comment]
@equanz
Thanks for checking.
I'm not sure why but the not-consuming issue can't be reproduced for some reason.
So now I can consume messages.

I also re-tested and still see NG.
Can you try several times ?
Sometimes it doesn't occur but it does most of the time.

**Comment by equanz** [comment]
Sorry to be late.

> I also re-tested and still see NG.
Can you try several times ?
Sometimes it doesn't occur but it does most of the time.

Try to check 7 times but I couldn't reproduce this issue.
I think https://github.com/apache/pulsar/issues/7455 possibly be another issue...

> Regarding the debug log, sorry I'm not sure how to enable debug log in Pulsar.

Here is an example. I thought try to check not-consuming issue by this log.
```
diff --git a/conf/log4j2.yaml b/conf/log4j2.yaml
index 34826740ca..b4dc4dabbb 100644
--- a/conf/log4j2.yaml
+++ b/conf/log4j2.yaml
@@ -151,6 +151,11 @@ Configuration:
         AppenderRef:
           - ref: Console

+      - name: org.apache.pulsar.broker.service.persistent.PersistentStickyKeyDispatcherMultipleConsumers
+        level: debug
+        AppenderRef:
+          - ref: Console
+
     # Logger to inject filter script
 #     - name: org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl
 #       level: debug
```

> I'm not sure why but the not-consuming issue can't be reproduced for some reason.

I understand. If this issue is caused by this issue, I must fix.

**Comment by feeblefakie** [comment]
@equanz Hmm, that is pretty weird.
It happens to me 9 out of 10.
The cause of #7455 seems like the number of consumers are changing (increasing) during consuming,
but is it the case for your testing ? 
(Sorry my procedures to reproduce is a bit not correct.
I added step 3 to register subscriptions beforehand.
https://github.com/feeblefakie/misc/blob/master/pulsar/HOW-TO-REPRODUCE.md )

If I start consumers before producers, the issue won't happen since the number of consumers is constant.

**Comment by equanz** [comment]
> The cause of #7455 seems like the number of consumers are changing (increasing) during consuming,
but is it the case for your testing ?

I think my issue is occurred by some consumers with different throughput in same subscription.

**Comment by sijie** [comment]
@codelipenghui can you take a look?

**Comment by codelipenghui** [comment]
@equanz Is it possible to add a unit test? so that we can better understand the problem and make sure it is fixed.

**Comment by equanz** [comment]
@codelipenghui I understand. As stated above, I think this issue is corner case and probably not easier to check.
But I'll try to add some tests.

**Comment by equanz** [comment]
@codelipenghui Added unit test to broker side. So, please take another look.

**Comment by codelipenghui** [comment]
@equanz Thanks a lot, I will take a look soon.

**Review by codelipenghui — APPROVED** [review]
